File: src/Canaan.Base/Logger.fs
namespace Canaan

open System

/// Logging interface
[<AbstractClass>]
type Logger() =
    static member val IsConfigured = false with get,set
    
    abstract member Info : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Debug : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Error : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Error : ex:Exception * messageTemplate:string * [<ParamArray>]args:obj[] -> unit

/// Basic console logger available anywhere
type ConsoleLogger() =
    inherit Logger()

    override this.Debug (messageTemplate:string, [<ParamArray>] args: obj[]) = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override this.Info (messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override this.Error (messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override this.Error (ex: Exception, messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"

File: src/Base/Canaan.Base/Logger.cs
using System;

namespace Canaan
{
    public abstract class Logger
    {
        public bool IsConfigured { get; protected set; } = false;

        public abstract void Info(string messageTemplate, params object[] args);

        public abstract void Debug(string messageTemplate, params object[] args);

        public abstract void Error(string messageTemplate, params object[] args);

        public abstract void Error(Exception ex, string messageTemplate, params object[] args);
    }

    public class ConsoleLogger : Logger
    {
        public override void Info(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Debug(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Error(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Error(Exception ex, string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);
    }
}
